src/contracts/AccessControlShares.sol

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import {MultiLP} from "./MultiLP.sol";
+
+abstract contract AccessControlShares is MultiLP {
+    uint256 public totalSupplyCap;
+    mapping(address lp => uint256 shares) public liquidityProviderCaps;
+
+    uint256[50] private _gap;
+
+    event LiquidityProviderCap(address indexed liquidityProvider, uint256 cap);
+    event TotalSupplyCap(uint256 shares);
+
+    function _postDepositHook(uint256) internal virtual override {
+        require(liquidityProviderCaps[msg.sender] >= balanceOf(msg.sender), "ARM: LP cap exceeded");
+        // total supply has already been updated
+        require(totalSupplyCap >= totalSupply(), "ARM: Supply cap exceeded");
+    }
+
+    function _postWithdrawHook(uint256) internal virtual override {
+        // Do nothing
+    }
+
+    function setLiquidityProviderCap(address liquidityProvider, uint256 cap) external onlyOwner {
```

Comment:
We are definitely going to want to make this one operate in batches. No need for the single version.

Reply:
good idea. I'll do today

Reply:
@DanielVF do we want to have multiple lp addresses and only one cap amount specified on setLiquidityProviderCap? Or the more flexible separate cap amounts for each LP?